### ISSUE TITLE
Using reserved root as nonce for mid-state caching in KMAC

### ISSUE THREAD

**Issue by kilpatty** [handshake-org/hsd#203]
Due to the fact that reserved root being 0s is not consensus, there seems to be an unintended optimization that a miner can make by abusing the reserved root as their nonce. 

In the proof of work function here: https://github.com/handshake-org/hsd/blob/master/lib/primitives/abstractblock.js#L286-L288

We can see that first the header is serialized to "prehead" data (All header values minus the nonce), and then fed into KMAC256 with the nonce as the key. 

When a miner is mining as intended, this function works well. The problem arises when a miner purposefully does not increment the key (nonce). If the key stays constant throughout mining, then a miner can theoretically take advantage of caching the state in the hashing algorithm (Keccack) - allowing them to save a decent amount of cycles.

Internally Keccack uses 136 byte blocks. The Handshake block header is 240 bytes. The nonce occupies the last 32 bytes, so the "prehead" data is 208 bytes. The reserved root occupies the space between byte 164-196. This means that the reserved root can be entirely changed without impacting the first block in Keccack, allowing that miner to cache that first block and not recalculate. 

When I attempted this exploit, I found that the KMAC operation does 4 block transforms. I looked at the state after each transform and found that the first 3 transforms stay entirely consistent while incrementing the reserved root instead of the nonce. This would allow a miner to cache 3 out of 4 operations done by KMAC saving them a significant amount of time and cycles. 

I think there are a number of ways for this to be solved, but I'm not sure which direction would be the best to take it. Hopefully this issue can serve as a starting point for discussion on how to fix this.

**Comment by pinheadmz** [comment]
Great find, I wasn't sure the reserved root could actually be used in a form of ASICboost.

Example of arbitrary data (ahem, possibly ASCII-encoded) as valid in reserved root:

https://hnscan.com/block/8058

**Comment by tynes** [comment]
Once the filter/reserved root *is* consensus, then this will no longer be a problem, but there is uncertainty around what to put in the roots, so it may be awhile before it becomes consensus. Once you commit to a certain things being placed in the roots, it would be a hardfork to change them. 

It seems like there are multiple solutions to this problem and we should use this thread to determine the pros and cons of each approach.

### Make null values for the reserved/filter root consensus

This is not ideal because it would require a hardfork to upgrade

### Move the reserved/filter root to different places in the block header, so that miners cannot benefit as much from asicboost like techniques.

We would need to determine the ideal layout of a block header in this case.

I'd also like to ask the question, allowing for asicboost is bad? It technically allows for more hashrate per dollar which I am under the assumption is good for the security of the network in general. If this optimization/exploit is not widely used, then it does create an unfair advantage for those that use it. This is why I believe asicboost like techniques are generally considered bad.

**Comment by chjj** [comment]
Seriously good catch. The last reserved root was added absent mindedly without realizing the impact it has on PoW. The filter root definitely affects it too, considering we probably won't have that ready for mainnet. Aside from the reserved roots, it looks like other malleable bits of data also contribute to this, including the version field. The solution here is to probably add _all_ of the miner-controlled data to the key (or the first block of keccak).

It looks like this optimization results in a pretty sizeable speedup (at least on an intel core i5). Rough estimates from keccak.org and blake2.net:

- SHA3-256 - [6.86 cpb][sha3] on a core i5
- BLAKE2b-256 - [3.08 cpb][blake2] on a core i5

Our headers are 240 bytes, but we don't include the nonce so it becomes 208 bytes.

```
(208 + 135) / 136 = 2 SHA3 blocks (+1 for the key)
(208 + 127) / 128 = 2 BLAKE2 blocks (+1 for the key)
```

The calculation becomes:

```
3 * (6.86 * 136) + 3 * (3.08 * 128) = 3981.6 cycles
```

Our PoW function is supposed to consume 3,981 cycles on an i5 6600.

With this optimization (abusing the version field):

```
2 * (6.86 * 136) + 3 * (3.08 * 128) = 3048.64 cycles
```

That's a hefty speedup, at least on consumer hardware. Though, this would be tricky since a miner may unintentionally end up voting on softfork bits.

I always thought satoshi could have prevented bitcoin's midstate optimization (along with other weird things) by using a SHA256 HMAC with the nonce as the key, but I hadn't considered that the version field would end up just becoming the nonce at that point. This would be especially true in the early days of bitcoin, since the block version was more or less meaningless.

But yeah, the bigger problem seems to be the reserved roots: if we don't use the reserved roots before mainnet (or allow them to be non-zero), a miner can also precompute the second block while grinding the roots in the third, resulting in a 2,115 cycle pow function:

```
1 * (6.86 * 136) + 3 * (3.08 * 128) = 2115.6800000000003
```

At that point the performance gains are nearly on par with bitcoin's midstate optimization. So, the construction we used to try to avoid this (KMAC), would do nothing (and actually seems to have made the problem worse).

So for a really simple solution, a new hash serialization should include the malleable fields in the first block of keccak-256 (136 byte block size):

```
version (4 bytes)
witness_root (32 bytes)
filter root (32 bytes)
reserved root (32 bytes)
nonce (32 bytes)
timestamp lo (4 bytes)
```

For a total of 136 bytes.

With the non-malleable fields following:

```
timestamp hi (4 bytes)
prev_block (32 bytes)
merkle_root (32 bytes)
urkel_root (32 bytes)
target (4 bytes)
```

For a total of 104 bytes (and a grand total of 240).

The timestamp overlaps with the second block by 4 bytes. The most significant bits are in the _second_ block as a result of it being little endian. I think this is okay for 3 reasons:

- The top 3 bytes will not be used for many thousands of years.
- The byte below that will not be used until after the year 2106.
- Even then, messing with the high bits will alter the time such that you would
  end up mining an invalid block, since you are altering the timestamp by
  years. The timestamp can only be safely fudged within the window of minutes
  usually (assuming `mtp <= timestamp <= adjusted time + 2 hours`), meaning
  only the low bits can be altered willy-nilly.

That said, once the timestamp reaches 5 bytes, every so often an increase in time carries all the way to the high bits. This appears to happen once every 136 years or so:

```
> new Date(0x0100000000 * 1000)
2106-02-07T06:28:16.000Z
> new Date(0x0200000000 * 1000)
2242-03-16T12:56:32.000Z
> new Date(0x0300000000 * 1000)
2378-04-22T19:24:48.000Z
```

This doesn't seem like an issue. There will be a minutes-long period every century-or-so where miners can try to game this on an extremely miniscule scale. Spookily enough, the number of years causing the carry is the exact block size of keccak-256 (136).

The txid merkle root is also miner-controlled technically, but the witness root being present in the first block already protects against any exploitation there.

Now that we have this covered, the call to blake2 should basically be ungame-able since the resulting keccak output acts as a new initialization vector for blake2b each time (no precomputation should be possible here).

Anyway, it goes without saying that we should probably drop the KMAC construction and switch to a naked round of SHA3. Ironically, the whole purpose of the using KMAC was to avoid optimizations like this, but I guess we didn't put _enough_ stuff in the key.

The final construction in full:

```
hdr = version
   || witness_root
   || filter_root
   || reserved_root
   || nonce
   || timestamp
   || prev_block
   || merkle_root
   || urkel_root
   || bits

blake2_key = SHA3_256(hdr)
pow_hash = BLAKE2b_256(hdr, blake2_key)
```

This will cause some minor rewrites for GPU code.

Thoughts? Maybe a new name? SHLAKE2b? BLA256? :)

[sha3]: https://keccak.team/sw_performance.html
[blake2]: https://blake2.net/

**Comment by smcki012** [comment]
Quick turn around for a fix. Seems straightforward and this was fantastic find @kilpatty. Big ups to @pinheadmz for pointing out this could potentially be used to exploit an optimization by showing us the reserved root could be altered. + 1 for BLA256.

**Comment by tynes** [comment]
I vote on being explicit and calling it SHA3-BLAKE2b256 instead of making up a new name that isn't particularly descriptive. SHA3-BLAKE for short.

**Comment by pinheadmz** [comment]
Based on https://github.com/handshake-org/hsd/commit/e916c35d59b464ff0f3145a04c2b1fb58a64190f

I think I've constructed an accurate diagram outlining the the new PoW data structure:
https://gist.githubusercontent.com/pinheadmz/7a54354f528d4db2f7dd09c3d75319ad/raw/c9b8710de204866580339f06b6f12629f8858b65/hsd_pow.txt

![Screen Shot 2019-07-11 at 12 43 12 PM](https://user-images.githubusercontent.com/2084648/61080237-78b6fe00-a3d9-11e9-93e9-2d3eb4ac232d.png)

@chjj Would love a review of the diagram to make sure we all understand what's going on ;-)

**Comment by chjj** [comment]
@pinheadmz that looks more or less correct.

The pow we came up with there is... complicated, but for good reasons.

The new design serves a few purposes. The first one is, instead of trying to avoid precomputation, we _encourage_ precomputation. There is a precomputation phase, and then the mining phase.

The precomputation not only allows us to fit everything into perfect blake2 and keccak blocks, but it also gives us the ability to "partially" merkelize the header, such that SPV resolvers don't have to sync as much data (they only need the preheader, not the subheader).

The next thing I discovered is related to timestamps. Adding a huge nonce to the header (in contrast to bitcoin, which has a small nonce) __destroys__ incentives for miners to update the timestamp. They will simply hash on the header forever while the timestamp grows stale. Bitcoin (possibly inadvertently) provides incentives for miners to keep the timestamp up-to-date by forcing miners to recompute the merkle tree when they run out of nonce space. A smart miner will update the timestamp whenever they can to avoid these several extra rounds of hashing.

We wanted to keep the big nonce in the header, but instead, it's been split into a small nonce and an extra nonce. In order to update the extra nonce, the subhead hash and commitment hash need to be recomputed (note that the subheader contains most of the malleable data). In other words, miners are _penalized_ for altering any of the malleable data, including the extra nonce.

To further increase incentives for timekeeping, we made it extra nice and placed the nonce _before_ the timestamp. This means that if it takes you a second or more to overflow the nonce, you can simply let the nonce overflow into the timestamp (treating the nonce and the first 4 bytes of the timestamp as a little-endian uint64). So, miners basically get an 8 byte nonce there.

Finally comes the XORing bit. This was inspired by an idea Kevin Pan posted to the bitcoin-dev ML. A mining pool can optionally commit to a 32 byte mask which is meant to be XOR'd with the header hash (I've been referring to the pre-xor'd hash as the "share hash" and the post-xor'd hash as the "pow hash"). The result of the XOR is compared against the network target. While this seems scary, remember that the miner has to _include it in the header_. If hash functions work the way we think they do (i.e. they're not broken), it should be impossible for anyone to game this, since they're XORing with unpredictable bytes.

The idea is that a mining pool can pre-commit to a mask that they keep secret. The pow function itself only commits to a hash of the mask, so there is no need to send the mask itself to individual miners. A typical mask would have bits set below the share target, but above the network target. This way, it becomes impossible for an individual miner to know whether they've found a block or not (they won't be able to compute the actual block hash).

I think this is a good thing to do in response to all of the block withholding attacks happening on other coins. The last thing I want is someone attacking this coin from the mining side.

I'm sure stratum will need to be revamped almost completely. We will have to figure that out as we go along I suppose.

**Comment by pinheadmz** [comment]
@chjj as long as you're buttoning up everything for the PoW, I wonder why a few small fields are still unchecked by consensus and currently filled with random bytes:

[Coinbase input[0] nSequence](https://github.com/handshake-org/hsd/blob/4ae81085e7365da4762572c2db52fd5e866fe316/lib/mining/template.js#L257)

[Coinbase witness stack item 1](https://github.com/handshake-org/hsd/blob/cc1ef7ab8dfcadbcc1b18b934ccf2d2ec1776042/lib/mining/template.js#L276) (item 0 is miner tagging and item 2 is `0x0000000000000000`)

Also, pow-ng removes the `filterRoot` but keeps the `reservedRoot` -- do you have a short list of options for the `reservedRoot`? I wonder if a UTXO commitment would be the best use of that space (h/t @bucko13). It seems to be something Bitcoin "wishes" it had, but may be too hard to insert after too much network life.

**Comment by kilpatty** [comment]
So just in spirit of considering every possible optimization, I have been mulling this over since working with @tynes on the mask testing yesterday. I'm not exactly sure if this can be used as an optimization, but I figured I can give an example of what I'm thinking. 

From my understanding most GPUs have no native types to understand a target or a hash in it's integer form (u256), therefore this comparison usually uses some kind of vector of u64s to compare and is somewhat expensive.

If a miner starts their next block knowing that the target is 007f01....(These bytes don't matter for the example). Which in binary form would be: 

0000 0000 0111 1111 0000 0001 

I could set my mask to have the same number of leading zeros, and then fill the remaining bits as 1s. 

0000 0000 0111 1111 1111 1111

I can now make the assumption while mining that if the hash that I find has the same number of leading zeros (or more), I'm (nearly) guaranteed to have found a nonce since any 1 in my hash beyond 9 0s would trigger a 0, and thus a smaller number. I believe the only case in which this would not return a valid nonce is if we matched the target exactly after XORing. 

e.g. 
Network Target =  0000 0000 0111 1111 0000 0001
Mask = 0000 0000 0111 1111 1111 1111
Share = 0000 0000 0101 1111 1111 1111
Result of XOR = 0000 0000 0010 000 0000 0000 < 0000 0000 0111 1111 0000 0001
so I have found a valid nonce. 

From my research most GPUs have built in support for this using the clz() function. I believe miners could optimize their miners by replacing a u256 comparison with a clz, and u32 comparison (clz(hash) >= num zeros target). 

That being said, in reality we only have to check the coefficient of the target as all other values will be 0. The coefficient (I believe) is only 3 bytes, so we can reduce that initial comparison from u256 < u256 to a u32 < u32 (And an operation to ensure the offset is similar). So I doubt that clz() will actually provide any increase if the CL code is being implemented correctly.

So all in all I don't know if this an optimization per se, but I don't know enough about ASICs/FPGAs/Hardware in general to know if this can't be abused in some manner. I figured by putting it out there someone who knows more might be able to tell if counting the leading zeros could ever be faster than a u32 comparison.

**Comment by chjj** [comment]
@kilpatty, I don't see how the mask affects this. This kind of comparison optimization would be possible on bitcoin. The real mask-related optimization a solo/farm miner would get is the ability to set their mask to all zeroes, which allows them to skip XORing altogether. We should assume that any solo/farm miner is using a mask of all zeroes. I think this is fine because the XORing isn't intended to be any kind of performance bottle-neck.

> I could set my mask to have the same number of leading zeros, and then fill the remaining bits as 1s.

I'm not sure what you mean here. Setting bits to 1s gets you nothing as you cannot predict the bits in the hash. All of the share hash bits will be unpredictable and distributed. Why wouldn't a miner just set it to zeroes, skip the XORing, and then do the comparison optimization?

**Comment by kilpatty** [comment]
@chjj I don't believe this optimization is possible in bitcoin (to it's fullest extent) as you can't make the same assumptions as you could here. E.g. (Assuming a 4 byte target for simplicity of the example. 

So after reading this through one more time, I think the true optimization comes from setting the exact same bits in the mask as the target. 

If we set the mask bits to the same bits in the target. We can change the operation from:
hash < target to "does clz(hash) exist in lookup table." Where the lookup table can be created before the initial hashing rounds start.

Before we start the hashing we find every 1 bit in the target and set the same bit in the mask. 
We then add the indexes of those bits to a shared memory lookup table on the GPUs. 

So for example:
```
0110 1001 - Target
0110 1001 - Mask
[1, 2, 4, 7] - Lookup Table
```

We then hash our block header, and perform a find first set (clz) on it. If that value exists in the lookup table, it's a valid nonce. 

Examples: 

```
0110 1001 - Target
0110 1001 - Mask
[1, 2, 4, 7] - Lookup Table
0010 1111 - Hash
0100 0110 - Result of XOR
clz(hash) == 2, which exists in table so valid nonce.
```

```
0110 1001 - Target
0110 1001 - Mask
[1, 2, 4, 7] - Lookup Table
1011 1111 - Hash
1101 0110 - Result of XOR
clz(hash) == 0, which is not in table, so invalid nonce.
```

```
0110 1001 - Target
0110 1001 - Mask
[1, 2, 4, 7] - Lookup Table
0000 0001 - Hash
0110 1000 - Result of XOR
clz(hash) == 7, which exists in table so valid nonce.
```

Technically this could be done in btc, but I think the operation would have to have 2 branches to it.
BTC:
```
if clz(hash) < clz(target) {
    valid nonce
} else if clz(hash) == clz(target) {
    need to do full comparison
} else {
    invalid nonce
}
```


I don't really think this can be taken advantage of but I thought it might be interesting to point out as I couldn't stop thinking about it after working on some of the mask stuff. Also let me know if I have some mistakes in the above, was getting confusing trying to keep it all in my head

**Comment by chjj** [comment]
The new PoW has been merged into master. Closing.
